File: webclient/src/sound/Tone.ts
import Packet from '#/io/Packet.js';

import Envelope from '#/sound/Envelope.js';

export default class Tone {
    static toneSrc: Int32Array | null = null;
    static noise: Int32Array | null = null;
    static sin: Int32Array | null = null;

    static tmpPhases: Int32Array = new Int32Array(5);
    static tmpDelays: Int32Array = new Int32Array(5);
    static tmpVolumes: Int32Array = new Int32Array(5);
    static tmpSemitones: Int32Array = new Int32Array(5);
    static tmpStarts: Int32Array = new Int32Array(5);

    frequencyBase: Envelope | null = null;
    amplitudeBase: Envelope | null = null;
    frequencyModRate: Envelope | null = null;
    frequencyModRange: Envelope | null = null;
    amplitudeModRate: Envelope | null = null;
    amplitudeModRange: Envelope | null = null;
    envRelease: Envelope | null = null;
    envAttack: Envelope | null = null;

    harmonicVolume: Int32Array = new Int32Array(5);
    harmonicSemitone: Int32Array = new Int32Array(5);
    harmonicDelay: Int32Array = new Int32Array(5);

    toneStart: number = 0;
    toneLength: number = 500;
    reverbVolume: number = 100;
    reverbDelay: number = 0;

    static init(): void {
        this.noise = new Int32Array(32768);
        for (let i: number = 0; i < 32768; i++) {
            if (Math.random() > 0.5) {
                this.noise[i] = 1;
            } else {
                this.noise[i] = -1;
            }
        }

        this.sin = new Int32Array(32768);
        for (let i: number = 0; i < 32768; i++) {
            this.sin[i] = (Math.sin(i / 5215.1903) * 16384.0) | 0;
        }

        this.toneSrc = new Int32Array(220500); // 10s * 22050 KHz
    }

    generate(sampleCount: number, length: number): Int32Array {
        for (let sample: number = 0; sample < sampleCount; sample++) {
            Tone.toneSrc![sample] = 0;
        }

        if (length < 10) {
            return Tone.toneSrc!;
        }

        const samplesPerStep: number = (sampleCount / length) | 0;

        this.frequencyBase?.reset();
        this.amplitudeBase?.reset();

        let frequencyStart: number = 0;
        let frequencyDuration: number = 0;
        let frequencyPhase: number = 0;

        if (this.frequencyModRate && this.frequencyModRange) {
            this.frequencyModRate.reset();
            this.frequencyModRange.reset();
            frequencyStart = (((this.frequencyModRate.end - this.frequencyModRate.start) * 32.768) / samplesPerStep) | 0;
            frequencyDuration = ((this.frequencyModRate.start * 32.768) / samplesPerStep) | 0;
        }

        let amplitudeStart: number = 0;
        let amplitudeDuration: number = 0;
        let amplitudePhase: number = 0;
        if (this.amplitudeModRate && this.amplitudeModRange) {
            this.amplitudeModRate.reset();
            this.amplitudeModRange.reset();
            amplitudeStart = (((this.amplitudeModRate.end - this.amplitudeModRate.start) * 32.768) / samplesPerStep) | 0;
            amplitudeDuration = ((this.amplitudeModRate.start * 32.768) / samplesPerStep) | 0;
        }

        for (let harmonic: number = 0; harmonic < 5; harmonic++) {
            if (this.frequencyBase && this.harmonicVolume[harmonic] !== 0) {
                Tone.tmpPhases[harmonic] = 0;
                Tone.tmpDelays[harmonic] = this.harmonicDelay[harmonic] * samplesPerStep;
                Tone.tmpVolumes[harmonic] = ((this.harmonicVolume[harmonic] << 14) / 100) | 0;
                Tone.tmpSemitones[harmonic] = (((this.frequencyBase.end - this.frequencyBase.start) * 32.768 * Math.pow(1.0057929410678534, this.harmonicSemitone[harmonic])) / samplesPerStep) | 0;
                Tone.tmpStarts[harmonic] = ((this.frequencyBase.start * 32.768) / samplesPerStep) | 0;
            }
        }

        if (this.frequencyBase && this.amplitudeBase) {
            for (let sample: number = 0; sample < sampleCount; sample++) {
                let frequency: number = this.frequencyBase.evaluateAt(sampleCount);
                let amplitude: number = this.amplitudeBase.evaluateAt(sampleCount);

                if (this.frequencyModRate && this.frequencyModRange) {
                    const rate: number = this.frequencyModRate.evaluateAt(sampleCount);
                    const range: number = this.frequencyModRange.evaluateAt(sampleCount);
                    frequency += this.generate2(range, frequencyPhase, this.frequencyModRate.form) >> 1;
                    frequencyPhase += ((rate * frequencyStart) >> 16) + frequencyDuration;
                }

                if (this.amplitudeModRate && this.amplitudeModRange) {
                    const rate: number = this.amplitudeModRate.evaluateAt(sampleCount);
                    const range: number = this.amplitudeModRange.evaluateAt(sampleCount);
                    amplitude = (amplitude * ((this.generate2(range, amplitudePhase, this.amplitudeModRate.form) >> 1) + 32768)) >> 15;
                    amplitudePhase += ((rate * amplitudeStart) >> 16) + amplitudeDuration;
                }

                for (let harmonic: number = 0; harmonic < 5; harmonic++) {
                    if (this.harmonicVolume[harmonic] !== 0) {
                        const position: number = sample + Tone.tmpDelays[harmonic];

                        if (position < sampleCount) {
                            Tone.toneSrc![position] += this.generate2((amplitude * Tone.tmpVolumes[harmonic]) >> 15, Tone.tmpPhases[harmonic], this.frequencyBase.form);
                            Tone.tmpPhases[harmonic] += ((frequency * Tone.tmpSemitones[harmonic]) >> 16) + Tone.tmpStarts[harmonic];
                        }
                    }
                }
            }
        }

        if (this.envRelease && this.envAttack) {
            this.envRelease.reset();
            this.envAttack.reset();

            let counter: number = 0;
            let muted: boolean = true;

            for (let sample: number = 0; sample < sampleCount; sample++) {
                const releaseValue: number = this.envRelease.evaluateAt(sampleCount);
                const attackValue: number = this.envAttack.evaluateAt(sampleCount);

                let threshold: number;
                if (muted) {
                    threshold = this.envRelease.start + (((this.envRelease.end - this.envRelease.start) * releaseValue) >> 8);
                } else {
                    threshold = this.envRelease.start + (((this.envRelease.end - this.envRelease.start) * attackValue) >> 8);
                }

                counter += 256;
                if (counter >= threshold) {
                    counter = 0;
                    muted = !muted;
                }

                if (muted) {
                    Tone.toneSrc![sample] = 0;
                }
            }
        }

        if (this.reverbDelay > 0 && this.reverbVolume > 0) {
            const start: number = this.reverbDelay * samplesPerStep;

            for (let sample: number = start; sample < sampleCount; sample++) {
                Tone.toneSrc![sample] += ((Tone.toneSrc![sample - start] * this.reverbVolume) / 100) | 0;
                Tone.toneSrc![sample] |= 0;
            }
        }

        for (let sample: number = 0; sample < sampleCount; sample++) {
            if (Tone.toneSrc![sample] < -32768) {
                Tone.toneSrc![sample] = -32768;
            }

            if (Tone.toneSrc![sample] > 32767) {
                Tone.toneSrc![sample] = 32767;
            }
        }

        return Tone.toneSrc!;
    }

    generate2(amplitude: number, phase: number, form: number): number {
        if (form === 1) {
            return (phase & 0x7fff) < 16384 ? amplitude : -amplitude;
        } else if (form === 2) {
            return (Tone.sin![phase & 0x7fff] * amplitude) >> 14;
        } else if (form === 3) {
            return (((phase & 0x7fff) * amplitude) >> 14) - amplitude;
        } else if (form === 4) {
            return Tone.noise![((phase / 2607) | 0) & 0x7fff] * amplitude;
        } else {
            return 0;
        }
    }

    read(dat: Packet): void {
        this.frequencyBase = new Envelope();
        this.frequencyBase.read(dat);

        this.amplitudeBase = new Envelope();
        this.amplitudeBase.read(dat);

        if (dat.g1() !== 0) {
            dat.pos--;

            this.frequencyModRate = new Envelope();
            this.frequencyModRate.read(dat);

            this.frequencyModRange = new Envelope();
            this.frequencyModRange.read(dat);
        }

        if (dat.g1() !== 0) {
            dat.pos--;

            this.amplitudeModRate = new Envelope();
            this.amplitudeModRate.read(dat);

            this.amplitudeModRange = new Envelope();
            this.amplitudeModRange.read(dat);
        }

        if (dat.g1() !== 0) {
            dat.pos--;

            this.envRelease = new Envelope();
            this.envRelease.read(dat);

            this.envAttack = new Envelope();
            this.envAttack.read(dat);
        }

        for (let harmonic: number = 0; harmonic < 10; harmonic++) {
            const volume: number = dat.gsmarts();
            if (volume === 0) {
                break;
            }

            this.harmonicVolume[harmonic] = volume;
            this.harmonicSemitone[harmonic] = dat.gsmart();
            this.harmonicDelay[harmonic] = dat.gsmarts();
        }

        this.reverbDelay = dat.gsmarts();
        this.reverbVolume = dat.gsmarts();
        this.toneLength = dat.g2();
        this.toneStart = dat.g2();
    }
}

File: javaclient/src/main/java/jagex2/sound/Tone.java
package jagex2.sound;

import deob.ObfuscatedName;
import jagex2.io.Packet;

@ObfuscatedName("dc")
public class Tone {

	@ObfuscatedName("dc.d")
	public Envelope frequencyBase;

	@ObfuscatedName("dc.e")
	public Envelope amplitudeBase;

	@ObfuscatedName("dc.f")
	public Envelope frequencyModRate;

	@ObfuscatedName("dc.g")
	public Envelope frequencyModRange;

	@ObfuscatedName("dc.h")
	public Envelope amplitudeModRate;

	@ObfuscatedName("dc.i")
	public Envelope amplitudeModRange;

	@ObfuscatedName("dc.j")
	public Envelope release;

	@ObfuscatedName("dc.k")
	public Envelope attack;

	@ObfuscatedName("dc.l")
	public int[] harmonicVolume = new int[5];

	@ObfuscatedName("dc.m")
	public int[] harmonicSemitone = new int[5];

	@ObfuscatedName("dc.n")
	public int[] harmonicDelay = new int[5];

	@ObfuscatedName("dc.o")
	public int reverbDelay;

	@ObfuscatedName("dc.p")
	public int reverbVolume = 100;

	@ObfuscatedName("dc.q")
	public int length = 500;

	@ObfuscatedName("dc.r")
	public int start;

	@ObfuscatedName("dc.s")
	public static int[] buf;

	@ObfuscatedName("dc.t")
	public static int[] noise;

	@ObfuscatedName("dc.u")
	public static int[] sine;

	@ObfuscatedName("dc.v")
	public static int[] fPos = new int[5];

	@ObfuscatedName("dc.w")
	public static int[] fDel = new int[5];

	@ObfuscatedName("dc.x")
	public static int[] fAmp = new int[5];

	@ObfuscatedName("dc.y")
	public static int[] fMulti = new int[5];

	@ObfuscatedName("dc.z")
	public static int[] fOffset = new int[5];

	@ObfuscatedName("dc.a()V")
	public static final void init() {
		noise = new int[32768];
		for (int var0 = 0; var0 < 32768; var0++) {
			if (Math.random() > 0.5D) {
				noise[var0] = 1;
			} else {
				noise[var0] = -1;
			}
		}
		sine = new int[32768];
		for (int var1 = 0; var1 < 32768; var1++) {
			sine[var1] = (int) (Math.sin((double) var1 / 5215.1903D) * 16384.0D);
		}
		buf = new int[220500];
	}

	@ObfuscatedName("dc.a(II)[I")
	public final int[] generate(int arg0, int arg1) {
		for (int var3 = 0; var3 < arg0; var3++) {
			buf[var3] = 0;
		}
		if (arg1 < 10) {
			return buf;
		}
		double var4 = (double) arg0 / ((double) arg1 + 0.0D);
		this.frequencyBase.genInit();
		this.amplitudeBase.genInit();
		int var6 = 0;
		int var7 = 0;
		int var8 = 0;
		if (this.frequencyModRate != null) {
			this.frequencyModRate.genInit();
			this.frequencyModRange.genInit();
			var6 = (int) ((double) (this.frequencyModRate.end - this.frequencyModRate.start) * 32.768D / var4);
			var7 = (int) ((double) this.frequencyModRate.start * 32.768D / var4);
		}
		int var9 = 0;
		int var10 = 0;
		int var11 = 0;
		if (this.amplitudeModRate != null) {
			this.amplitudeModRate.genInit();
			this.amplitudeModRange.genInit();
			var9 = (int) ((double) (this.amplitudeModRate.end - this.amplitudeModRate.start) * 32.768D / var4);
			var10 = (int) ((double) this.amplitudeModRate.start * 32.768D / var4);
		}
		for (int var12 = 0; var12 < 5; var12++) {
			if (this.harmonicVolume[var12] != 0) {
				fPos[var12] = 0;
				fDel[var12] = (int) ((double) this.harmonicDelay[var12] * var4);
				fAmp[var12] = (this.harmonicVolume[var12] << 14) / 100;
				fMulti[var12] = (int) ((double) (this.frequencyBase.end - this.frequencyBase.start) * 32.768D * Math.pow(1.0057929410678534D, (double) this.harmonicSemitone[var12]) / var4);
				fOffset[var12] = (int) ((double) this.frequencyBase.start * 32.768D / var4);
			}
		}
		for (int var13 = 0; var13 < arg0; var13++) {
			int var14 = this.frequencyBase.genNext(arg0);
			int var15 = this.amplitudeBase.genNext(arg0);
			if (this.frequencyModRate != null) {
				int var16 = this.frequencyModRate.genNext(arg0);
				int var17 = this.frequencyModRange.genNext(arg0);
				var14 += this.waveFunc(var17, this.frequencyModRate.form, var8) >> 1;
				var8 += (var16 * var6 >> 16) + var7;
			}
			if (this.amplitudeModRate != null) {
				int var18 = this.amplitudeModRate.genNext(arg0);
				int var19 = this.amplitudeModRange.genNext(arg0);
				var15 = var15 * ((this.waveFunc(var19, this.amplitudeModRate.form, var11) >> 1) + 32768) >> 15;
				var11 += (var18 * var9 >> 16) + var10;
			}
			for (int var20 = 0; var20 < 5; var20++) {
				if (this.harmonicVolume[var20] != 0) {
					int var21 = var13 + fDel[var20];
					if (var21 < arg0) {
						buf[var21] += this.waveFunc(var15 * fAmp[var20] >> 15, this.frequencyBase.form, fPos[var20]);
						fPos[var20] += (var14 * fMulti[var20] >> 16) + fOffset[var20];
					}
				}
			}
		}
		if (this.release != null) {
			this.release.genInit();
			this.attack.genInit();
			int var22 = 0;
			boolean var23 = false;
			boolean var24 = true;
			for (int var25 = 0; var25 < arg0; var25++) {
				int var26 = this.release.genNext(arg0);
				int var27 = this.attack.genNext(arg0);
				int var28;
				if (var24) {
					var28 = this.release.start + ((this.release.end - this.release.start) * var26 >> 8);
				} else {
					var28 = this.release.start + ((this.release.end - this.release.start) * var27 >> 8);
				}
				var22 += 256;
				if (var22 >= var28) {
					var22 = 0;
					var24 = !var24;
				}
				if (var24) {
					buf[var25] = 0;
				}
			}
		}
		if (this.reverbDelay > 0 && this.reverbVolume > 0) {
			int var29 = (int) ((double) this.reverbDelay * var4);
			for (int var30 = var29; var30 < arg0; var30++) {
				buf[var30] += buf[var30 - var29] * this.reverbVolume / 100;
			}
		}
		for (int var31 = 0; var31 < arg0; var31++) {
			if (buf[var31] < -32768) {
				buf[var31] = -32768;
			}
			if (buf[var31] > 32767) {
				buf[var31] = 32767;
			}
		}
		return buf;
	}

	@ObfuscatedName("dc.a(IIII)I")
	public final int waveFunc(int arg1, int arg2, int arg3) {
		if (arg2 == 1) {
			return (arg3 & 0x7FFF) < 16384 ? arg1 : -arg1;
		} else if (arg2 == 2) {
			return sine[arg3 & 0x7FFF] * arg1 >> 14;
		} else if (arg2 == 3) {
			return ((arg3 & 0x7FFF) * arg1 >> 14) - arg1;
		} else if (arg2 == 4) {
			return noise[arg3 / 2607 & 0x7FFF] * arg1;
		} else {
			return 0;
		}
	}

	@ObfuscatedName("dc.a(ZLmb;)V")
	public final void unpack(Packet arg1) {
		this.frequencyBase = new Envelope();
		this.frequencyBase.unpack(arg1);
		this.amplitudeBase = new Envelope();
		this.amplitudeBase.unpack(arg1);
		int var3 = arg1.g1();
		if (var3 != 0) {
			arg1.pos--;
			this.frequencyModRate = new Envelope();
			this.frequencyModRate.unpack(arg1);
			this.frequencyModRange = new Envelope();
			this.frequencyModRange.unpack(arg1);
		}
		int var4 = arg1.g1();
		if (var4 != 0) {
			arg1.pos--;
			this.amplitudeModRate = new Envelope();
			this.amplitudeModRate.unpack(arg1);
			this.amplitudeModRange = new Envelope();
			this.amplitudeModRange.unpack(arg1);
		}
		int var5 = arg1.g1();
		if (var5 != 0) {
			arg1.pos--;
			this.release = new Envelope();
			this.release.unpack(arg1);
			this.attack = new Envelope();
			this.attack.unpack(arg1);
		}
		for (int var6 = 0; var6 < 10; var6++) {
			int var7 = arg1.gsmarts();
			if (var7 == 0) {
				break;
			}
			this.harmonicVolume[var6] = var7;
			this.harmonicSemitone[var6] = arg1.gsmart();
			this.harmonicDelay[var6] = arg1.gsmarts();
		}
		this.reverbDelay = arg1.gsmarts();
		this.reverbVolume = arg1.gsmarts();
		this.length = arg1.g2();
		this.start = arg1.g2();
	}
}
